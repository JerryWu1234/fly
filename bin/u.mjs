#!/usr/bin/env node
'use strict'
import { Ufn, runCli } from '../dist/index.js'
runCli(Ufn)
