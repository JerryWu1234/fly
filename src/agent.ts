export type Agent = 'bun' | 'pnpm' | 'yarn' | 'npm'

export const LOCKS: Record<string, Agent> = {
  'yarn.lock': 'yarn',
  'pnpm-lock.yaml': 'pnpm',
  'package-lock.json': 'npm',
  'bun.lockb': 'bun',

}

export const INIT = {
  yarn: 'yarn',
  npm: 'npm install',
  pnpm: 'pnpm install',
  bun: 'bun install',
}

export const AGENT = {
  a: {
    yarn: 'yarn add',
    npm: 'npm install',
    pnpm: 'pnpm install',
    bun: 'bun install',
  },
  r: {
    yarn: 'yarn run',
    npm: 'npm run',
    pnpm: 'pnpm run',
    bun: 'bun run',
  },
  u: {
    yarn: 'yarn remove',
    npm: 'npm uninstall',
    pnpm: 'pnpm remove',
    bun: 'bun remove',
  },
}

