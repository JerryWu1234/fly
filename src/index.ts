export { Afn } from './command/a'
export { Ufn } from './command/u'
export * from './run'
