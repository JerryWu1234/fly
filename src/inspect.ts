import { findUp } from 'find-up'
export async function inspect() {
  
}
