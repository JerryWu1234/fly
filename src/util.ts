import { execSync } from 'child_process'
import os from 'os'

export function detectRepeatCmd(agent: string) {
  try {
    execSync(os.platform() === 'win32'
      ? `cmd /c "(help ${agent} > nul || exit 0) && where ${agent} > nul 2> nul"`
      : `command -v ${agent}`)
    return true
  }
  catch (e) {
    return false
  }
}
