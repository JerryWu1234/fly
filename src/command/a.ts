import type { Agent } from '../agent'
import { AGENT, INIT } from '../agent'
import type { Fn } from '../type'
const NAME = 'a'
const AGENT_NAME = AGENT[NAME]

export const Afn: Fn = (agent: Agent, args: string[]) => {
  if (args.length === 0)
    return INIT[agent]

  const get_agent = AGENT_NAME.install[agent]
  return `${get_agent} ${args.join(' ')}`
}

