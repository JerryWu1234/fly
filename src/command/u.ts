import type { Agent } from '../agent'
import { AGENT } from '../agent'
import type { Fn } from '../type'
const NAME = 'u'
const AGENT_NAME = AGENT[NAME]

export const Ufn: Fn = (agent: Agent, args: string[]) => {
  if (args.length === 0) {
    console.error('please input command')
    process.exit(1)
  }

  const get_agent = AGENT_NAME[agent]
  return `${get_agent} ${args.join(' ')}`
}

