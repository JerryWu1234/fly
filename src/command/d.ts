import type { Agent } from '../agent'
import { AGENT } from '../agent'
import type { Fn } from '../type'
import { getPackageJson } from '../util'
import { getStorage, setStorage } from '../storage'
const NAME = 'r'
const AGENT_NAME = AGENT[NAME]

export const Dfn: Fn = (agent: Agent, args: string[], cwd?: string) => {
  const storage = getStorage()
  const json = getPackageJson(cwd as string)
  if (args.length === 0 && !storage[json.name]) {
    console.error('please input command')
    process.exit(1)
  }
  else {
    const res = recordCommand(agent, args, cwd as string, storage)
    setStorage(res)
    return res[json.name]
  }

  return getExecuteCmd(agent, args)
}

export function recordCommand(agent: Agent, args: string[], cwd: string, mergeObject: Record<string, any>) {
  const json = getPackageJson(cwd)
  let returnVal: Record<string, any> = {}
  if (mergeObject[json.name] && args.length === 0) {
    return mergeObject
  }
  else {
    returnVal[json.name] = getExecuteCmd(agent, args)
    if (Object.keys(mergeObject))
      returnVal = Object.assign(mergeObject, returnVal)
    return returnVal
  }
}

function getExecuteCmd(agent: Agent, args: string[]) {
  const get_agent = AGENT_NAME[agent]
  return `${get_agent} ${args.join(' ')}`
}
