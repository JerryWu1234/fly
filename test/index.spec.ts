import { expect, test } from 'vitest'
import { inspect } from '../src/inspect'
import { Afn, Ufn, run } from '../src/index'
import { getPackageJson } from '../src/util'
import { recordCommand } from '../src/command/u'
test('inspect', async () => {
  const cwdA = `${process.cwd()}/feature/a`
  const resultA = await inspect({ cwd: cwdA })
  expect(resultA).toBe('pnpm')
  const cwdB = `${process.cwd()}/feature/b`
  const resultB = await inspect({ cwd: cwdB })
  expect(resultB).toBe(null)
  const cwd = `${process.cwd()}`.slice(0, -5)
  const result = await inspect({ cwd })
  expect(result).toBe('pnpm')
})

test('test a command', async () => {
  process.env.NODEDEV = 'debug'
  const resultA = await run(Afn, [], `${process.cwd()}/feature/a`)
  expect(resultA).toBe('pnpm install')
})

test('test Afn', async () => {
  const val = Afn('yarn', ['install', '-g'])
  expect(val, 'yarn add install -g')
})

test('test Ufn', async () => {
  const val = Ufn('yarn', ['dev'])
  expect(val, 'yarn run dev')
})

test('test read package.json', async () => {
  const cwdA = `${process.cwd()}/feature/a`
  const jsonA = getPackageJson(cwdA)
  expect(jsonA).toMatchSnapshot()
  const cwdB = `${process.cwd()}/feature/a/pnpm-lock.yaml`
  const jsonB = getPackageJson(cwdB)
  expect(jsonB).toMatchSnapshot()
})

test('test merge object', () => {
  const cwdA = `${process.cwd()}/feature/a`
  const v = recordCommand('pnpm', ['dev', '--port:300'], cwdA, {})
  expect(v).toMatchSnapshot()
  const b = recordCommand('pnpm', ['dev', '--port:300'], cwdA, {
    feature: 'pnpm run dev --port:300',
    aa: 'pnpm run dev',
  })
  expect(b).toMatchSnapshot()
  const c = recordCommand('pnpm', ['dev', '--port:300'], cwdA, {
    aa: 'pnpm run dev',
  })
  expect(c).toMatchSnapshot()
})
