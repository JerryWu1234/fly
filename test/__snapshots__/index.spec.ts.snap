// Vitest Snapshot v1

exports[`test merge object 1`] = `
{
  "feature": "pnpm run dev --port:300",
}
`;

exports[`test merge object 2`] = `
{
  "aa": "pnpm run dev",
  "feature": "pnpm run dev --port:300",
}
`;

exports[`test merge object 3`] = `
{
  "aa": "pnpm run dev",
  "feature": "pnpm run dev --port:300",
}
`;

exports[`test read package.json 1`] = `
{
  "author": "",
  "description": "",
  "keywords": [],
  "license": "ISC",
  "main": "index.js",
  "name": "feature",
  "scripts": {
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
  "version": "1.0.0",
}
`;

exports[`test read package.json 2`] = `
{
  "author": "",
  "description": "",
  "keywords": [],
  "license": "ISC",
  "main": "index.js",
  "name": "feature",
  "scripts": {
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
  "version": "1.0.0",
}
`;
